KEP-3926: Add tests for storage list error aggregation and surface the abort reason in list errors

tkashem opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Behavior change: Previously the abort fired on the error after the aggregator was full, so a list with exactly 100 corrupt items completed the scan and returned 100 errors + `CauseTypeTooMany`. Now it aborts at the 100th, drops any items after it from the result, and reports CauseTypeTooMany. So the change limits the length of the error slice to 100. 

This PR adds a unit test for storage `GetList` to verify that list error aggregation works as expected

#### Which issue(s) this PR fixes:
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
KEP: https://github.com/kubernetes/enhancements/pull/3927
```

tkashem committed:
refactor: move 'ListErrorAggregator' interface to storage

Jefftree commented:
/triage accepted

tkashem commented:
stress test result:
```
❯ go test -race k8s.io/apiserver/pkg/storage/etcd3 -c
❯ stress -p=64 ./etcd3.test -test.run=TestGetListWithErrorAggregation
5s: 0 runs so far, 0 failures
10s: 0 runs so far, 0 failures
15s: 5 runs so far, 0 failures
20s: 64 runs so far, 0 failures
...
7m0s: 1643 runs so far, 0 failures
7m5s: 1666 runs so far, 0 failures
7m10s: 1682 runs so far, 0 failures
```

tkashem commented:
/cc @dgrisonnet

tkashem commented on staging/src/k8s.io/apiserver/pkg/storage/testing/store_tests.go:
```diff
@@ -1687,6 +1690,330 @@ func RunTestGetListRecursivePrefix(ctx context.Context, t *testing.T, store stor
+			verifier: func(t *testing.T, aggr *FakeListErrorAggregator, list *example.PodList, err error) {
+				// a) aggregation should stop after the max limit of 100 is reached
+				// and the 101th entry should be the 'too many' sentinel error
+				limit := 100
```
i was looking at the other side of the coin, if someone changes the value accidentally then the test will fail and will work as an alarm, so the test holds the spec independently; if both the test and the implementation share the constant then the test will alway quietly pass irrespective of whether the change is intentional.

tkashem commented on staging/src/k8s.io/apiserver/pkg/storage/testing/store_tests.go:
```diff
@@ -1687,6 +1690,330 @@ func RunTestGetListRecursivePrefix(ctx context.Context, t *testing.T, store stor
+		{
+			name:           "feature enabled, first error is unexpected, no aggregation expected",
+			featureEnabled: true,
+			// objects numbered foo-{2|4|6} will become corrupt,
```
fixed

tkashem commented:
/cc @stlaz @sttts

tkashem committed:
add tests to verify error aggregation for storage GetList

ibihim commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/corrupt_obj_deleter.go:
```diff
@@ -82,14 +82,14 @@ type corruptObjErrAggregator struct {
-	}
 	var corruptObjErr *corruptObjectError
 	if errors.As(err, &corruptObjErr) {
 		a.errs = append(a.errs, storage.NewCorruptObjError(key, corruptObjErr))
```
Resolved, thx

ibihim commented:
/remove-lifecycle rotten

tkashem committed:
rename interface to ListItemErrors

tkashem commented on staging/src/k8s.io/apiserver/pkg/storage/testing/store_tests.go:
```diff
@@ -2839,6 +2840,444 @@ func RunTestGetListRecursivePrefix(ctx context.Context, t *testing.T, store stor
+			//  a) GetList encounters an unexpected (not corruptObjErr)
+			//  error while retrieving {2} and immediately aborts
+			//  b) GetList successfully decodes {1}
+			verifier: func(t *testing.T, recorder *listErrorAggregatorRecorder, list *example.PodList, err error) {
```
i have a test case that proves that in the default case (for backward compatibility) the err is returned as is (identical to the one created by the List operation). we need the recorder for that, no?

ibihim replied:
You are both right, the recorder did not prove anything from the outside. Dropped it, the tests now assert only on the error / list returned by `GetList`. 1f30426

everettraven commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/corrupt_obj_deleter.go:
```diff
@@ -81,15 +81,19 @@ type corruptObjErrAggregator struct {
+
 	var corruptObjErr *corruptObjectError
 	if errors.As(err, &corruptObjErr) {
 		a.errs = append(a.errs, storage.NewCorruptObjError(key, corruptObjErr))
+		if len(a.errs) >= a.maxCount {
+			// add a sentinel error to indicate there are more
+			a.errs = append(a.errs, errTooMany)
+			return true
+		}
```
Minor thing, but because we are actually appending the original error to the list and _then_ checking if it is `>= maxCount` before appending the sentinel error the sentinel error ends up being the `maxCount+1`th error.

The GoDoc on the `corruptObjErrAggregatorFactory` function signals that `maxCount` is the maximum errors to return, so it may be a bit surprising to caller implementations when they always receive `maxCount+1` errors if they only read the description of the function instead of looking through the implementation.

I don't have a preference as to whether or not the GoDoc or the implementation is updated, but making sure they match up would be a good idea IMO.

ibihim replied:
Good point @everettraven. Went with @benluddy's route, `errTooMany` is the `abortErr` now instead of a `maxCount+1`th element, so `a.errs` holds exactly `maxCount` and the double length check is gone. de9dd89

everettraven reviewed (commented):
I'm not all that familiar with the broader KEP work, but @ibihim had asked me to take a look at this PR.

Non-authoritative review from me. Most of my comments are related to the test implementation which I find more difficult to follow than I suspect is necessary, but this could just be personal preference.

benluddy commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/corrupt_obj_deleter.go:
Shouldn't `abortErr` be surfaced somehow when `len(a.errs) > 0 && a.abortErr != nil`?

ibihim replied:
Good catch. `Aggregate()` carries `abortErr` alongside the collected errs now. de9dd89

benluddy commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:
```diff
@@ -190,6 +190,23 @@ func TestKeySchema(t *testing.T) {
 
+func TestGetListWithErrorAggregation(t *testing.T) {
+	storagetesting.RunTestGetListWithErrorAggregation(t, func(t *testing.T) (context.Context, *storagetesting.ErrorAggregatorFactory, storagetesting.InterfaceWithCorruptTransformer) {
+		ctx, s, _ := testSetup(t)
```
Was this inversion necessary because the shared test code is overriding the feature gate?

ibihim replied:
Removed the inversion, the caller passes the store directly now instead of the `*T`→factory closure. 5561c4e

ibihim committed:
remove recorder, use black-box assertions only

Remove listErrorAggregatorRecorder and ErrorAggregatorFactory from
GetList error aggregation tests. Verifiers now assert solely on the
error and list returned by GetList instead of inspecting aggregator
internals.

ibihim committed:
fix: surface abortErr in aggregated storage errors

Move errTooMany from a.errs slice to a.abortErr, fixing the off-by-one
where a.errs contained maxCount+1 entries. Pass abortErr through to
aggregatedStorageError so it is surfaced as a StatusCause, fixing the
bug where unexpected abort errors were silently dropped when corrupt
object errors had already been collected.

ibihim committed:
move feature gate toggling from shared test code to caller

The shared storagetesting code was toggling AllowUnsafeMalformedObjectDeletion
internally. Convention is that callers toggle feature gates, not shared helpers.
This also eliminates the fragile pattern where the newStoreFn closure reads
gate state that was set inside the shared code.

- Extract backward-compat test case into RunTestGetListBackwardCompatibility
- Remove featureEnabled field and SetFeatureGateDuringTest from shared code
- Split etcd3 caller into two test functions, each setting the gate upfront

ibihim committed:
remove factory inversion from shared error aggregation tests

Pass (ctx, t, store) directly to RunTestGetListWithErrorAggregation
and RunTestGetListWithoutErrorAggregation, matching the convention used
by other shared storage tests. Each subtest now uses a unique namespace
to isolate objects within a shared store, eliminating the need for a
factory callback.

Rename RunTestGetListBackwardCompatibility to RunTestGetListWithoutErrorAggregation
and TestGetListBackwardCompatibility to TestGetListWithoutErrorAggregation.

ibihim committed:
decouple shared test fixtures from etcd3 transformer implementation

Move the test transformer and marker constants from etcd3/store_test.go
into storagetesting, so the shared test code owns the full fixture
(which objects to mark, how the transformer fails). The etcd3 side now
only provides the UpdateTransformer mechanism and the corruptObjectError
value that the aggregator recognizes.

ibihim commented on staging/src/k8s.io/apiserver/pkg/storage/interfaces.go:
```diff
@@ -420,3 +420,18 @@ func PrepareKey(resourcePrefix, key string, recursive bool) (string, error) {
+	// err: it represents the error the list operation encountered while
+	// retrieving the given object from the storage.
+	// done: true if the aggregation is done and the list operation should
+	// abort, otherwise the list operation will continue
```
Renamed to `ListItemErrors` and took your godoc almost verbatim. The n→1 confusion should be gone now. 288a241

ibihim commented on staging/src/k8s.io/apiserver/pkg/storage/interfaces.go:
```diff
@@ -420,3 +420,18 @@ func PrepareKey(resourcePrefix, key string, recursive bool) (string, error) {
+	// retrieving the given object from the storage.
+	// done: true if the aggregation is done and the list operation should
+	// abort, otherwise the list operation will continue
+	Aggregate(key string, err error) bool
```
Went with `Append(key, err) bool`. 288a241

ibihim commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:
```diff
@@ -190,6 +190,23 @@ func TestKeySchema(t *testing.T) {
+	storagetesting.RunTestGetListWithErrorAggregation(t, func(t *testing.T) (context.Context, *storagetesting.ErrorAggregatorFactory, storagetesting.InterfaceWithCorruptTransformer) {
+		ctx, s, _ := testSetup(t)
+		var store storage.Interface = s
+		if utilfeature.DefaultFeatureGate.Enabled(features.AllowUnsafeMalformedObjectDeletion) {
```
Moved the gate toggling to the caller (`TestGetListWithErrorAggregation` / `TestGetListWithoutErrorAggregation`), the shared code no longer touches it. 5adfaa5

ibihim commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:
```diff
@@ -210,20 +227,32 @@ type corruptedTransformer struct {
+		return out, stale, &corruptObjectError{err: fmt.Errorf("bits flipped"), errType: untransformable}
+	case strings.Contains(string(data), storagetesting.UnexpectedErrKey):
+		return out, stale, fmt.Errorf("bits flipped")
+	}
```
Yes. Added `UpdateTransformer(TransformerModifier)`, the shared test sets its own transformer now and the etcd3 coupling is gone. 109e213

ibihim commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:
```diff
@@ -191,17 +191,22 @@ func TestKeySchema(t *testing.T) {
 	})
 }
 
+func TestGetListBackwardCompatibility(t *testing.T) {
```
Done, `TestGetListWithoutErrorAggregation`. 5561c4e

ibihim committed:
storage/testing: move away from annotation based failure to funcs

ibihim commented on staging/src/k8s.io/apiserver/pkg/storage/testing/store_tests.go:
```diff
@@ -2839,6 +2840,444 @@ func RunTestGetListRecursivePrefix(ctx context.Context, t *testing.T, store stor
+		// the function decides whether the object represented
+		// by the given id should be marked to become corrupt or
+		// fail to transform with an unexpected error
+		corrupter func(id int) string
```
I pushed back on this at first, but gave it a try, and it did improve the code. The annotation markers are gone. Each case now hands the transformer an `errorInjector` (`func(key string) error`), so "which object fails, and how" lives in the test. Done in 34fb26a1dfb. Thanks.

ibihim commented on staging/src/k8s.io/apiserver/pkg/storage/testing/store_tests.go:
```diff
@@ -2839,6 +2840,444 @@ func RunTestGetListRecursivePrefix(ctx context.Context, t *testing.T, store stor
+		// recorder: records the values GetList passes to the error aggregator
+		// list: result of GetList operation is saved into list
+		// err: error returned from GetList
+		verifier func(t *testing.T, _ *listErrorAggregatorRecorder, list *example.PodList, err error)
```
Gave this one a try too, but the change wasn't worth it. Too much variability between the cases.

benluddy commented:
/remove-lifecycle rotten

jpbetz commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/corrupt_obj_deleter.go:
```diff
@@ -81,15 +81,19 @@ type corruptObjErrAggregator struct {
 	var corruptObjErr *corruptObjectError
 	if errors.As(err, &corruptObjErr) {
 		a.errs = append(a.errs, storage.NewCorruptObjError(key, corruptObjErr))
+		if len(a.errs) >= a.maxCount {
+			// treat exceeding the limit as an abort condition
+			a.abortErr = errTooMany
+			return true
```
This changes behavior.

Previously the abort fired on the error *after* the aggregator was full, so a list with exactly 100 corrupt items completed the scan and returned 100 errors. Now it aborts at the 100th, drops any items after it from the result, and reports `CauseTypeTooMany`.

This looks like a reasonable improvement, but I wasn't expecting to find this in a PR titled "Add a test to...".

Would you please update the title and description.  This helps both with review, and helps document the purpose and impact of this PR.

ibihim replied:
Addressed the title in https://github.com/kubernetes/kubernetes/pull/129129#issuecomment-4989340501. I can't easily change the description though.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/corrupt_obj_deleter.go:
```diff
@@ -217,6 +221,9 @@ func (e *corruptObjectError) Error() string {
 	errs           utilerrors.Aggregate
+	// abortErr is the reason the aggregation stopped, if any.
+	// It is nil when all items were processed without hitting a limit or unexpected error.
+	abortErr error
```
Since `errTooMany` now lives here instead of inside `errs`, `aggregatedStorageError.Error()` no longer mentions the too-many condition in the error string (e.g. in the V(5) log). It only shows up in the API status causes. Is this intentional?

ibihim replied:
I will update the `Error()` output to contain the hint of `too many`.

ibihim replied:
Added as suffix with https://github.com/kubernetes/kubernetes/pull/129129/changes/6387daaf22a015f0b2f12610bb4f6fa310cc0087

jpbetz commented on staging/src/k8s.io/apiserver/pkg/storage/interfaces.go:
```diff
@@ -425,3 +425,17 @@ func PrepareKey(resourcePrefix, key string, recursive bool) (string, error) {
+// each Append, it signals whether it considers itself full or the error to be fatal.
+// In either case, the caller is supposed to not keep collecting, but return the
+// collection.
+type ListItemErrors interface {
```
```suggestion
type listItemErrors interface {
```

Recommend not exporting unless there is a compelling reason.

ibihim replied:
It was exported for the shared test fixtures in storage/testing, which this PR decoupled. Moved it back into etcd3 and unexported it: https://github.com/kubernetes/kubernetes/pull/129129/changes/e4f89dcaf7e8cbb16f6a85736e4fb56906fa94a8

jpbetz commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:
```diff
@@ -276,25 +291,19 @@ func (s *storeWithPrefixTransformer) UpdatePrefixTransformer(modifier storagetes
+func (s *storeWithTransformerOverride) UpdateTransformer(modifier storagetesting.TransformerModifier) func() {
+	orig := s.store.transformer
+	s.store.transformer = modifier(orig)
+	s.store.watcher.transformer = modifier(orig)
```
```suggestion
	next := modifier(orig)
	s.store.transformer = next
	s.store.watcher.transformer = next
```

?  I can't think of a reason why it's better to do this twice.

ibihim replied:
Oh, this is pretty obviously bad. Fixed in https://github.com/kubernetes/kubernetes/pull/129129/changes/836214fffc8980e3e25daaa0b12f56b17ee05cd1

jpbetz commented on staging/src/k8s.io/apiserver/pkg/storage/testing/store_tests.go:
```diff
@@ -2802,6 +2802,422 @@ func RunTestGetListRecursivePrefix(ctx context.Context, t *testing.T, store stor
+					Name:      objNameFn(j),
+					Namespace: ns,
+				}}
+				_, _ = testPropagateStore(ctx, t, store, obj)
```
nit (fine as a follow-up): with n=210 this ends up doing a few hundred serial Delete+Create round-trips — the Delete inside `testPropagateStore` is always a no-op on these fresh namespaces. `RunTestListPaginationRareObject` just calls `store.Create` in a loop for its 1000-object seed; the same pattern would shave some time here.

ibihim replied:
Right, there was never anything to delete here (fresh store before, fresh namespaces now). Fixed in https://github.com/kubernetes/kubernetes/pull/129129/changes/d55e1923c2c40c22e25efc21be337d8260d46ff7

jpbetz commented on staging/src/k8s.io/apiserver/pkg/storage/testing/store_tests.go:
```diff
@@ -2802,6 +2802,422 @@ func RunTestGetListRecursivePrefix(ctx context.Context, t *testing.T, store stor
+			//  until it reaches the maximum limit, and then it immediately aborts
+			//  b) GetList successfully decodes the first 100 objects in the good set into list
+			verifier: func(t *testing.T, _ string, list *example.PodList, err error) {
+				limit := 100
```
nit (follow-up is fine): the 100 here mirrors the bare literal in the `corruptObjErrAggregatorFactory(100)` call on the etcd3 side. A shared exported constant would keep them from drifting — though at least the failure would be loud if they did.

ibihim replied:
hm, I think having 100 as a const is cleaner. So I added it as a const. Not eager though to share it across packages, which I can still do. Done in https://github.com/kubernetes/kubernetes/pull/129129/changes/c1f927dd01e6263f12e528b332b15b5dc7c7cb24.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/storage/testing/store_tests.go:
```diff
@@ -2802,6 +2802,422 @@ func RunTestGetListRecursivePrefix(ctx context.Context, t *testing.T, store stor
+// idOf extracts the numeric suffix from an object key like ".../foo-000002" -> 2.
+func idOf(key string) int {
+	name := key[strings.LastIndex(key, "/")+1:]
+	n, _ := strconv.Atoi(strings.TrimPrefix(name, "foo-"))
```
nit: dropping the `Atoi` error means any key that doesn't match `foo-<digits>` parses to id 0, which the even-id injectors then treat as corrupt. All keys match today, so it's latent — but a `t.Fatalf` on parse failure would make future reuse of these helpers safer.

ibihim replied:
safe is safe. I will add it.

https://github.com/kubernetes/kubernetes/pull/129129/changes/8db8bf1ec116c612315dfec3a59a6eef1f8ef935

jpbetz commented:
cc @serathius 

/hold
To discuss behavior changes. I was expecting this to only impact test, but the `TooMany` firing at the limit change is worthy of a title/description change and careful review and consideration.

ibihim commented:
/retitle KEP-3926: Add tests for storage list error aggregation and surface the abort reason in list errors

ibihim committed:
etcd3: surface the abort reason in the aggregated error message

Otherwise the error string logged at V(5) no longer mentions that the
list was truncated.

ibihim committed:
etcd3: move the listItemErrors interface back in and unexport it

Nothing outside etcd3 references it anymore.

ibihim committed:
etcd3: apply the transformer modifier once in UpdateTransformer

ibihim committed:
storage/testing: seed error aggregation tests with Create

The Delete inside testPropagateStore is always a no-op here, the
objects are created in fresh namespaces.

ibihim committed:
storage/testing: fail on malformed keys in idOf

A key that does not match foo-<digits> used to parse to id 0, which
the even-id injectors silently treat as corrupt.

ibihim committed:
etcd3: make the corrupt object error aggregation limit a const

ibihim commented:
/test pull-kubernetes-node-e2e-containerd

jpbetz reviewed (commented):
/approve
/lgtm

jpbetz commented:
/hold cancel

ibihim commented:
/test pull-kubernetes-integration
